### DISCUSSION TITLE
Column Compression

### DISCUSSION THREAD

**Discussion by TheAlexCH** [ClickHouse/ClickHouse#33606] [Q&A] [answered]
Hi,

I have a couple of tables that have hashes and I would like to know whats the maximum compression and which one would this be in order to save disk space?

Does this have an impact on performance, given that none of those columns is a sorting key?
(I know from RDBMS that compression is a trade off between CPU and disk/ram usage)

Also I wonder how clickhouse compression works. Its not live is that correct? because whenever I run imports, first the table allocate a lot of space and it then gradually drops over the next minutes.

**Comment by UnamedRus** [comment, marked as answer]
>I have a couple of tables that have hashes and I would like to know whats the maximum compression and which one would this be in order to save disk space?

How do you store those hashes? (as hex string?)
If so, you can use unhex/hex functions and FixedString datatype in order to encode them in 2 times less bytes.

For couple biggest columns in table, you can look into ZSTD codec, but it's slower on decompression.
https://kb.altinity.com/altinity-kb-schema-design/codecs/codecs-speed/
https://kb.altinity.com/altinity-kb-schema-design/codecs/altinity-kb-how-to-test-different-compression-codecs/

> Also I wonder how clickhouse compression works. Its not live is that correct? because whenever I run imports, first the table allocate a lot of space and it then gradually drops over the next minutes.

It's live.

It's because ClickHouse merge parts in background, and those parts being removed after 8 minutes of inactivity. 

It's controlled by merge_tree setting `old_parts_lifetime` default value set to 560 sec.

You can use those queries to find current disk usage by tables. (notice active=1, it's because we want to select only active parts)

https://kb.altinity.com/altinity-kb-useful-queries/altinity-kb-database-size-table-column-size/

**Comment by TheAlexCH** [comment]
yes they are stored as hex string. will check that out. Thanks a lot

Ok so basically the compression is live but clickhouse only starts removing merged parts after 8 minutes. why 8 minuted? why not 1? just wondering

**Reply by UnamedRus** [reply]
Because ClickHouse doesn't use fsync for writes.

So in case of hard restart and disappearing result of latest merge, ClickHouse will have source parts available to do merge of data again.
